xbout/region.py

```diff
@@ -0,0 +1,520 @@
+        # No upper X-point
+        if jys11 <= 0 and jys22 >= ny - 1:
+            ix = min(ixs1, ixs2)
+            if ix >= nx - 1:
+                return 'core'
+            elif ix <= 0:
+                return 'sol'
+            else:
+                return 'limiter'
+
+        return 'single-null'
+
+    if jys11 == jys21 and jys12 == jys22:
+        if jys11 < nyinner - 1 and jys22 > nyinner:
+            return 'xpoint'
+        else:
+            raise ValueError('Currently unsupported topology')
+
+    if ixs1 == ixs2:
+        if jys21 < nyinner - 1 and jys12 > nyinner:
+            return 'connected-double-null'
+        else:
+            raise ValueError('Currently unsupported topology')
+
+    return 'disconnected-double-null'
+
+
+def _create_connection_x(regions, inner, outer):
+    regions[inner].connection_outer = outer
+    regions[outer].connection_inner = inner
+
+
+def _create_connection_y(regions, lower, upper):
+    regions[lower].connection_upper = upper
+    regions[upper].connection_lower = lower
+
+
+def _create_regions_toroidal(ds):
+    topology = _get_topology(ds)
+
+    coordinates = {'t': ds.metadata.get('bout_tdim', None),
+                   'x': ds.metadata.get('bout_xdim', None),
+                   'y': ds.metadata.get('bout_ydim', None),
+                   'z': ds.metadata.get('bout_zdim', None)}
+
+    ixs1 = ds.metadata['ixseps1']
+    ixs2 = ds.metadata['ixseps2']
+    nx = ds.metadata['nx']
+
+    jys11 = ds.metadata['jyseps1_1']
+    jys21 = ds.metadata['jyseps2_1']
+    nyinner = ds.metadata['ny_inner']
+    jys12 = ds.metadata['jyseps1_2']
+    jys22 = ds.metadata['jyseps2_2']
+    ny = ds.metadata['ny']
+
+    mxg = ds.metadata['MXG']
+    myg = ds.metadata['MYG']
+    # keep_yboundaries is 1 if there are y-boundaries and 0 if there are not
+    ybndry = ds.metadata['keep_yboundaries']*myg
+    if jys21 == jys12:
+        # No upper targets
+        ybndry_upper = 0
+    else:
+        ybndry_upper = ybndry
+
+    # Make sure all sizes are sensible
+    ixs1 = _in_range(ixs1, 0, nx)
+    ixs2 = _in_range(ixs2, 0, nx)
+    ixs1, ixs2 = _order_vars(ixs1, ixs2)
+    jys11 = _in_range(jys11, 0, ny - 1)
+    jys21 = _in_range(jys21, 0, ny - 1)
+    jys12 = _in_range(jys12, 0, ny - 1)
+    jys21, jys12 = _order_vars(jys21, jys12)
+    nyinner = _in_range(nyinner, jys21 + 1, jys12 + 1)
+    jys22 = _in_range(jys22, 0, ny - 1)
+
+    # Adjust for boundary cells
+    # keep_xboundaries is 1 if there are x-boundaries and 0 if there are not
+    if not ds.metadata['keep_xboundaries']:
+        ixs1 -= mxg
+        ixs2 -= mxg
+        nx -= 2*mxg
+    jys11 += ybndry
+    jys21 += ybndry
+    nyinner += ybndry + ybndry_upper
+    jys12 += ybndry + 2*ybndry_upper
+    jys22 += ybndry + 2*ybndry_upper
+    ny += 2*ybndry + 2*ybndry_upper
+
+    # Note, include guard cells in the created regions, fill them later
+    regions = OrderedDict()
+    if topology == 'disconnected-double-null':
```

Comment:
Agree, but I'll update in parallel-interpolation branch because I ended up having to refactor the topology-defining bit due to needing the connections already in the Region constructor.